### ISSUE TITLE
Alternator: support new table attribute "TableClass"

### ISSUE THREAD

**Issue by nyh** [scylladb/scylla#10431]
DynamoDB [announced](https://aws.amazon.com/blogs/aws/new-dynamodb-table-class-save-up-to-60-in-your-dynamodb-costs/) in December 2021 a new attribute of a table, called _table class_. The choice of "table class" doesn't change anything in the table's capabilities - it just controls the way it is stored: There are currently two options ([described here](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/WorkingWithTables.tableclasses.html)) - STANDARD and STANDARD_INFREQUENT_ACCESS. The former is the usual tables DynamoDB always had. The latter is, as it's name suggests, meant for _infrequently-accessed_ tables, and its pricing is different: the data storage costs are lower (the announcement suggested 60% lower) but reads and writes are more expensive.

We can't support these specific table classes right now, although @psarna is working on various storage options so eventually we may be able to support these and other table classes. But we should start by supporting the new API (and possibly complain if the user is trying to use something other than the STANDARD table class):

1. `CreateTable` and `UpdateTable` support a new optional `TableClass` request attribute and returns a new `TableClassSummary` attribute in the response.
2. `DescribeTable` also returns a `TableClassSummary`.

**Comment by nyh** [comment]
@avikivity speculated that perhaps `STANDARD_INFREQUENT_ACCESS` doesn't use a different storage mechanism or disk type, but merely **enables compression** (or switches to a stronger and slower type of compression). It is reasonable that just enabling compression could lower storage requirement by 60% on average (Amazon could charge by some average expectation) while making reads and writes more expensive.

Whether or not this is what Amazon are doing in their `STANDARD_INFREQUENT_ACCESS`, it is definitely some something that we can easily support right now.

**Comment by mykaul** [comment]
I assume LZ4 -> ZSTD and ZSTD -> higher level of ZSTD compression level?

**Comment by nyh** [comment]
> I assume LZ4 -> ZSTD and ZSTD -> higher level of ZSTD compression level?

I see that there is no "compression_parameters" set on the Alternator table, and if I follow the code correctly (it's confusing...), this in turn means that lz4 is used by default. I'll create a test that writes some data to check that sstables are really compressed, and also check if zstd makes a noticable size reduction. If it does, we may want to support TableClass as a convenient and standard way to choose it.

UPDATE: the test in https://github.com/scylladb/scylladb/pull/20520 verifies that today, new Alternator tables do get LZ4Compressor enabled by default:
```
'compression': '({sstable_compression : org.apache.cassandra.io.compress.LZ4Compressor})'
```
So I do know for sure that Alternator uses LZ4.

**Comment by mykaul** [comment]
> > I assume LZ4 -> ZSTD and ZSTD -> higher level of ZSTD compression level?
> 
> I see that there is no "compression_parameters" set on the Alternator table, and if I follow the code correctly (it's confusing...), this in turn means that lz4 is used by default. I'll create a test that writes some data to check that sstables are really compressed, and also check if zstd makes a noticable size reduction. If it does, we may want to support TableClass as a convenient and standard way to choose it.

Hmmm. now that I think about it , what about our (upcoming) object storage? That's a perfect candidate for infrequent access...

**Comment by nyh** [comment]
According to the examples that @kbr-scylla gave in https://github.com/scylladb/scylladb/pull/4742#issuecomment-514629334 zstd gives some nice improvement over lz4, but NOT a 60% improvement, so it is more likely that in Amazon's implementation the difference is between no compression at all and yes compression - and not by two different types of compression, but this is just a guess (as I said, Avi made this guess, and it makes a lot of sense to me).

That being said, since this API exists, it does make sense to implement it using zstd. Even if it will give some users 20% less storage and not 60%, it can be helpful for certain users.

**Comment by nyh** [comment]
In the draft PR https://github.com/scylladb/scylladb/pull/20520 I create a full test for the TableClass API feature - that checks how this option is handled in CreateTable, UpdateTable and DescribeTable (all these tests pass on DynamoDB, fail on Scylla) plus checking that the TableClass controls the compression type. 

Now all that's needed is to actually implement this feature, not just the test :-) Should be fairly easy, shouldn't take more than a day. And then we would need to do benchmarks to see if switching from lz4 to zstd would really give a "typical" user a non-trivial stroage reduction at the cost of performance - if it's true it will be a good fit for TableClass.

**Comment by mykaul** [comment]
> In the draft PR #20520 I create a full test for the TableClass API feature - that checks how this option is handled in CreateTable, UpdateTable and DescribeTable (all these tests pass on DynamoDB, fail on Scylla) plus checking that the TableClass controls the compression type.
> 
> Now all that's needed is to actually implement this feature, not just the test :-) Should be fairly easy, shouldn't take more than a day. And then we would need to do benchmarks to see if switching from lz4 to zstd would really give a "typical" user a non-trivial stroage reduction at the cost of performance - if it's true it will be a good fit for TableClass.

Q: when switching, we do not re-compress existing data, right? 
Q2: please refer to https://github.com/scylladb/scylladb/issues/10431#issuecomment-2341029058

**Comment by nyh** [comment]
> Q: when switching, we do not re-compress existing data, right?

DynamoDB does allow you to use UpdateTable to **change** a table's TableClass. This is a background operation, which you can check when completed - and when it's completed you know you're paying the new prices (it's not documented what are the costs in the interim period). So whatever DynamoDB do internally (compression, different disks, etc.) they do guarantee to move the existing data and not just handle new data differently. DynamoDB only allows you to do these movements twice a day - presumably because the movement itself isn't charged and they don't want people to go crazy and force them to process your entire database for free 20 times a day.

We would eventually need to implement this movement (if compression is used - some sort of compaction to re-compress the data), but we can implement this in steps, with each step being useful:
1. Only allowing to pick tableclass at CreateTable time, 
2. Allow changing tableclass with UpdateTable, but it will only effect new compactions and won't re-compress existing sstables.
3. UpdateTable will start a background re-compression process (basically compaction).

>  Q2: please refer to [#10431 (comment)](https://github.com/scylladb/scylladb/issues/10431#issuecomment-2341029058)

Here you asked maybe we can use an object-storage backend to implement the infrequent-access table class. Definitely - in fact if you look at my original description of this option, this is what I thought we would do. However, Avi made a good point that looking at the price differences (60% discount on storage, just 25% markup on requests) and the performance difference (Amazon claims there is none) it really does sound like perhaps Amazon only changes compression and doesn't really change the storage to use slower disks or anything like that. But this is just a guess, and using better compression was just me hoping to get a quick win for users who complain about too much storage being used - without months of development of new storage backend, new sstable index formats, and so on.

**Comment by mykaul** [comment]
I'm not against the compression change. I'm just trying to be future-proof - when we WILL introduce object storage as a lower tier, how would Alternator users be able to specify it? Changing later the semantics seems strange to me.

**Comment by nyh** [comment]
> I'm not against the compression change. I'm just trying to be future-proof - when we WILL introduce object storage as a lower tier, how would Alternator users be able to specify it? Changing later the semantics seems strange to me.

I agree, and I don't know. I think that the best answer is that there is no reason why we need to limit ourselves to the two classes "STANDARD" and "STANDARD_WITH_INFREQUENT_ACCESS" that Amazon supports - we can add other ones with different names, e.g., "OBJECT_STORAGE" and the good thing is that the SDK will allow it (since those names are just strings, not keywords that the Python library needs to support).

We can also be "future proof" by not supporting Amazon's name "STANDARD_WITH_INFREQUENT_ACCESS" right now (we'll only support that only when we're 100% sure what we want it to be) - and for now only supporting a new table class name, e.g., "ZSTD_COMPRESSION". This won't help completely-unmodified applications (who don't use the new table class name which we invent), but it allows to easily modify the application to use ZSTD using official DynamoDB API. This is in contrast with what an application needs to do right now to change the compression (i.e., run ugly CQL commands after creating the table).
